libraries/state_history/log.cpp

```diff
@@ -37,12 +37,48 @@ void state_history_log_data::construct_index(const fc::path& index_file_name) co
 }
 
 state_history_log::state_history_log(const char* const name, const state_history_config& config)
-    : name(name) {
+    : name(name), logger(config.logger ? *config.logger : fc::logger::get(DEFAULT_LOGGER)) {
    catalog.open(config.log_dir, config.retained_dir, config.archive_dir, name);
    catalog.max_retained_files = config.max_retained_files;
    this->stride               = config.stride;
    open_log(config.log_dir / (std::string(name) + ".log"));
    open_index(config.log_dir / (std::string(name) + ".index"));
+
+   if (config.threaded_write) {
+      num_buffered_entries = config.num_buffered_entries;
+      thr = std::thread([this] {
+         fc_ilog(logger,"${name} thread created", ("name", this->name));
+         try {
+            while (!ending.load()) {
+               std::unique_lock lock(mx);
+               cv.wait(lock);
+               write_queue_type tmp;
+               std::swap(tmp, write_queue);
+               for (const auto& entry : tmp) {
+                  write_entry(entry, lock);
+               }
+            }
+         }
+         catch(...) {
+            fc_elog(logger,"catched exception from ${name} write thread", ("name", this->name));
+            eptr = std::current_exception();
+         }
+         fc_ilog(logger,"${name} thread ended", ("name", this->name));
+      });
+   }
+}
+
+void state_history_log::stop() {
+   if (thr.joinable()) {
+      ending = true;
+      cv.notify_one();
```

Comment:
It feels like there might be a race here -- `notify_one()` isn't guaranteed to be called after `wait(lock)` I believe?

More generally though, instead of rolling our own thread safe queue, it seems like using asio io_context would be better suited here. That's the pattern used throughout the rest of the eosio code.